### DISCUSSION TITLE
Suggested Design for communication between UI and Nebula Graph

### DISCUSSION THREAD

**Discussion by xs2tarunkukreja** [vesoft-inc/nebula#5591] [General]
Hi All,

We are using Nebula Graph having N types of entities and M types of relationship. We are planning to display the same on SPA(Angular). Currently I have developed an REST API for communication between UI and Graph DB.

I have developed endpoints to get entities and relationships. But while working on this approach, I realized this is not a sustainable solution.

Do any of you know any suggested design for the same? Any reference would be helpful.

Regards,
Tarun

**Comment by wey-gu** [comment]
Dear @xs2tarunkukreja 

Have you looked into this project? https://github.com/vesoft-inc/nebula-http-gateway NebulaGraph Studio/Explorer is actually built on top of that.

**Reply by xs2tarunkukreja** [reply]
Thanks. 

But I think there is a challenge. Here we are sending the query from UI to the gateway. Is it not a security concern like exposing the DB structure to user if UI is implemented as SPA?

If yes, then how to take care of the same?

Also, there are issue like SQL injection in SQL DB. Is there any such concern or security issue with Nebula Graph? If yes then how to take care of the same.

**Reply by wey-gu** [reply]
True, efforts on different layers(are still needed) as we cannot expose this directly to SPA.

If you consider your work is general enough, would you consider creating a lib as an opensource project? Thus it could in the future have joint efforts from the community.
